Time Zone in Account Settings

kevinjkim opened the pull request:
Added time zone field to user account settings. Time zone is stored in preferences rather than the student model.

### [TNL-4594](https://openedx.atlassian.net/browse/TNL-4594)

Below is what it looks like:
![d6dc6e3e-2817-11e6-89fb-17785a03399d](https://cloud.githubusercontent.com/assets/14193378/15826225/bf796c68-2bd4-11e6-9a9a-fb6365c00471.png)


### Sandbox
- [x] Build a sandbox for your branch and add a link here [https://kevinjkim.sandbox.edx.org/](https://kevinjkim.sandbox.edx.org/)

### Testing
- [ ] i18n
- [x] safecommit shows 0 violations
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @jcdyer 
- [x] Code review: @nasthagiri 
- [ ] UI strings/error msgs review: @catong 
- [ ] UX review: @chris-mike 
- [x] Accessibility review: @cptvitamin 
- [ ] Product review: @sstack22 

### Post-review
- [x] Squash commits

kevinjkim commented:
Currently, I display all pytz common_timezones (there's 431 of them). It looks kinda messy. Should I filter down based on the country that the user has selected ( and just display all if no country is selected )?

Also, there are a couple Antarctica timezones. Should I just omit those? Not sure how many people would use those.

nasthagiri commented:
I'm always a proponent of making the UX great for the user.  So if there's any intelligence we can add in minimizing the options, we should do it.  But only if we don't inadvertently prevent them from doing what they want.  Is there a reliable way to filter down the timezone options based on country that will always work?  Alternatively, instead of filtering the options, we can present the country-specific timezones on the top of the list - so it's easier for them to find.

edX is global.  Who are we to say you can't learn in Antartica?  :)

Thoughts @marcotuts and Shelby?

kevinjkim commented:
@catong Yes! Here is the new layout:

![screen shot 2016-06-06 at 1 42 30 pm](https://cloud.githubusercontent.com/assets/14193378/15831657/a9dbcaf0-2bec-11e6-86be-618d2ba69dec.png)

catong commented on lms/static/js/student_account/views/account_settings_factory.js:
```diff
@@ -100,6 +100,20 @@
+                                valueAttribute: 'time_zone',
+                                helpMessage: gettext(
+                                    'The time zone you currently reside in. ' +
+                                    'This is used to calculate course start/end times.'
```
I think it's more than just course start and end times. Perhaps say "... course dates" in keeping with the description in account settings?

kevinjkim replied:
Sorry. I haven't pushed the new code yet. That old code is for the first image. I updated it to look like the second image. I was just waiting to see if there were any more thoughts on list filtering so that i could push it all at once.

sstack22 commented:
@nasthagiri, @kevinjkim - I think we should list relevant time zones at the top of the list, and the remainder should be in alphabetical order. Here's an example that I found helpful: http://www.timeanddate.com/worldclock/converter.html

Not all learners set country field. Is it possible to get relevant time zones based user's IP address?

Any insights on why there are 341 options? I believe there are only ~32 time zones.

jcdyer commented:
@sstack22 There are more timezones than 32, because there is more information captured than just how far from UTC you are.  Daylight savings starts earlier in the US than Mexico, for instance, so they are represented by different timezones.  There are also locations which changed timezones at different points historically, and they have to be represented separately as well.

jcdyer commented:
@sstack22:  The PR does use `pytz.common_timezones`, which cuts out a lot of the historical timezones, but generally, each country has its own timezones, because they are politically administered, and subject to arbitrary changes.  

tl;dr time is complicated.

jcdyer commented:
Can you make sure that we validate that incoming timezones are legit timezones?  (Use `pytz.common_timezones_set` instead of `pytz.common_timezones` for validation, because testing for membership in a long list is potentially slow.)

Also make sure there's a test that shows the failure when an invalid timezone is submitted.

kevinjkim commented:
Yes. I just added a validation/test against incorrect timezones.

Also i moved the time zone field below the country field, based on feedback.

<img width="1219" alt="screen shot 2016-06-07 at 11 15 46 am" src="https://cloud.githubusercontent.com/assets/14193378/15863593/9724ad7c-2ca1-11e6-9786-a5f089242c72.png">

jcdyer commented on AUTHORS:
```diff
@@ -271,3 +271,4 @@ Dmitry Viskov <dmitry.viskov@webenterprise.ru>
 Brian Jacobel <bjacobel@edx.org>
 Sigberto Alarcon <salarcon@stanford.edu>
 Sofiya Semenova <ssemenova@edx.org>
+Kevin Kim <kkim@edx.org>
```
Nice.

jcdyer commented:
Once the feature flag work merges back in: :+1:

catong commented:
@sstack22 @kevinjkim @jcdyer @nasthagiri I've seen time zone lists where there's some relative offset from GMT, etc., which gives users some degree of confidence that they're picking the correct time zone. With the text-only/location name only list, is there some other means in the UI for learners to be confident that they've selected and are seeing the time they expected?

sstack22 commented:
@catong - We actually discussed the same thing earlier yesterday, and they are looking into it! Great suggestion :)

kevinjkim commented:
Just a look at how the offset looks in the dropdown now:

<img width="1167" alt="screen shot 2016-06-13 at 2 06 15 pm" src="https://cloud.githubusercontent.com/assets/14193378/16017914/1ca56b44-3170-11e6-976a-61dce0799f55.png">

catong commented:
@kevinjkim Thanks for the screenshot - that's always helpful. How is the "offset from" timezone determined for each country/region? e.g. Africa/Maputo is offset from CAT while Africa/Maseru is offset from SAST.
Not to introduce new churn if this has already been discussed, but was there any thought given to  making every time zone show offset from a single other timezone e.g. UTC? (especially since course dates and time show in UTC if local time zone is not specified. @sstack22?

sstack22 commented:
@catong - The offset is based on UTC. CAT, as an example, is that local time zone. That time zone is 2 hours ahead of UTC.

catong commented:
@sstack22 Do you mean, if I see "Africa/Maputo (CAT+0200)" that means the Africa/Maputo timezone is called "CAT" and is 2 hours ahead of UTC?

sstack22 commented:
Yes @catong  - @kevinjkim - can you update to reflect Carol's suggestion above? Happy to chat offline if helpful.

nasthagiri commented:
I find this syntax confusing.  I had the same interpretation as @catong.  Can we change it to something like, "(CAT, UTC+0200)".  So the timezone abbreviation is separated from the UTC offset.

jcdyer commented on common/djangoapps/util/date_utils.py:
```diff
@@ -5,7 +5,7 @@
 import re
 
-from pytz import timezone, UTC, UnknownTimeZoneError
+from pytz import timezone, UTC, utc, UnknownTimeZoneError
```
Standardize all uses on lower-case utc, and remove the UTC import.

jcdyer commented on common/djangoapps/util/date_utils.py:
```diff
@@ -62,6 +62,14 @@ def get_time_display(dtime, format_string=None, coerce_tz=None):
+    """
+    Returns a formatted time zone (e.g. 'Asia/Tokyo (JST +0900)') for user account settings time zone drop down
+    """
+    offset = utc.localize(datetime.utcnow()).astimezone(timezone(time_zone)).strftime('%Z%z')
```
Consider replacing `pytz.utc.localize(datetime.utcnow())` with [`django.utils.timezone.now()`](https://docs.djangoproject.com/en/1.9/ref/utils/#django.utils.timezone.now).

jcdyer commented on lms/djangoapps/student_account/test/test_views.py:
```diff
@@ -19,11 +19,13 @@
+from course_modes.models import CourseMode
 from openedx.core.djangoapps.user_api.accounts.api import activate_account, create_account
 from openedx.core.djangoapps.user_api.accounts import EMAIL_MAX_LENGTH
 from openedx.core.djangolib.js_utils import dump_js_escaped_json
```
Nit: Import order is a little wonky here.  stdlib first, then external libraries, then stuff in edx-platform.  One blank line between each group. `nose` should come in the same group as `django`.  Each group should be alphabetical.

jcdyer commented on lms/djangoapps/student_account/views.py:
```diff
@@ -47,6 +47,7 @@
 from openedx.core.djangoapps.user_api.errors import UserNotFound
 from util.date_utils import strftime_localized
 
+from openedx.core.djangoapps.user_api.models import UserPreference
```
Import order....

jcdyer commented on lms/static/js/student_account/views/account_settings_factory.js:
```diff
@@ -109,6 +109,22 @@
+                        {
+                            view: new AccountSettingsFieldViews.DropdownFieldView({
+                                model: userPreferencesModel,
+                                required: true,
```
Does this need to be required?

kevinjkim replied:
In the dropdown, there's a blank option initially when the user hasn't selected anything. Setting `required: true` gets rid of the blank option once the users selects a time zone.

jcdyer commented on lms/static/js/student_profile/views/learner_profile_fields.js:
```diff
@@ -16,22 +17,31 @@
+                var accountSettingsLink = HtmlUtils.joinHtml(
+                    HtmlUtils.interpolateHtml(
+                        HtmlUtils.HTML('<a href="{settings_url}">', {settings_url: this.options.accountSettingsPageUrl})
+                    ),
```
Does this function do the right thing WRT escaping accountSettingsPageUrl?

andy-armstrong replied:
I think the brackets are in the wrong place, as the interpolation parameters are incorrectly being passed to ``HtmlUtils.HTML``.

kevinjkim replied:
Sorry. Are you talking about the `{settings_url}` brackets?

jcdyer commented on openedx/core/djangoapps/user_api/models.py:
```diff
@@ -27,6 +30,10 @@ class UserPreference(models.Model):
     value = models.TextField()
 
+    TIME_ZONE_CHOICES = [
+        [tz, get_formatted_time_zone(tz)] for tz in common_timezones
```
Nit: Use a tuple for this inner pair (parentheses instead of square brackets).

jcdyer replied:
Important: Sort this list by formatted name (Don't be like Coursera!)

kevinjkim replied:
`get_formatted_time_zone()` keeps the time zone name at the beginning of the string so it should be alphabetical since `common_timezones` is sorted alphabetically

jcdyer replied:
True.  So this code is correct, but I'd still like it changed, because it's not robust.  If we decide to change the way we display timezones, this could easily become unusable without anyone noticing.  Including the sort function makes it more resilient to UX changes.

It won't cause a performance issue, because the sorting will only happen once, when the class is loaded.

jcdyer commented on openedx/core/djangoapps/user_api/preferences/api.py:
```diff
@@ -392,6 +394,18 @@ def validate_user_preference_serializer(serializer, preference_key, preference_v
         })
+    if preference_key == "time_zone" and preference_value not in common_timezones_set:
+        developer_message = ugettext_noop(u"Value '{preference_value}' not valid for preference '{preference_key}': "
+                                          u"Not in timezone set.")
```
I think you need to make this a single string literal, and mark it with `# pylint: disable=line-too-long` so the translation works properly.

jcdyer commented:
Pass complete.

andy-armstrong commented on lms/static/js/student_profile/views/learner_profile_fields.js:
```diff
@@ -16,22 +17,31 @@
-                        {'account_settings_page_link': accountSettingsLink}
-                    ));
+                    this._super(
+                        HtmlUtils.interpolateHtml(
```
This should use ``StringUtils.interpolate`` as it isn't creating HTML.

kevinjkim replied:
I just tried with `StringUtils.interpolate()` and that didn't work

andy-armstrong replied:
Oh, my mistake. I see that the link is being passed in, so there is HTML in line. Sorry for the confusion.

andy-armstrong commented on lms/static/js/student_profile/views/learner_profile_fields.js:
```diff
@@ -16,22 +17,31 @@
-                        {'account_settings_page_link': accountSettingsLink}
-                    ));
+                    this._super(
+                        HtmlUtils.interpolateHtml(
```
This should use ``StringUtils.interpolate`` as it isn't creating HTML.

nasthagiri commented on AUTHORS:
```diff
@@ -272,4 +272,4 @@ Brian Jacobel <bjacobel@edx.org>
 Sofiya Semenova <ssemenova@edx.org>
 Alisan Tang <atang@edx.org>
-
+Kevin Kim <kkim@edx.org>
```
Yay!

nasthagiri commented on common/djangoapps/util/date_utils.py:
```diff
@@ -62,6 +63,19 @@ def get_time_display(dtime, format_string=None, coerce_tz=None):
+    """
+    abbr = now().astimezone(timezone(time_zone)).strftime('%Z')
+    offset = now().astimezone(timezone(time_zone)).strftime('%z')
+    if option == 'abbr':
```
Why are these options passed in as strings instead of python keyword arguments?  Also, I don't see any calls that actually pass in a value for `option`.

kevinjkim replied:
I'll change it to use keyword arguments. Just realized none of the code here uses the `option` value but some of my other code for converting dates used it so I guess I got mixed up.

nasthagiri commented on common/djangoapps/util/date_utils.py:
```diff
@@ -62,6 +63,19 @@ def get_time_display(dtime, format_string=None, coerce_tz=None):
+        return abbr
+    elif option == 'offset':
+        return offset
+    return "{name} ({abbr}, UTC{offset})".format(name=time_zone, abbr=abbr, offset=offset).replace("_", " ")
```
Can you explain why replacing `_` is needed?

kevinjkim replied:
Can't remember who, but someone said they'd prefer not seeing the `_`s. So `America/New_York (EDT, UTC-400)` would be `America/New York (EDT, UTC-400)`

nasthagiri commented on common/djangoapps/util/date_utils.py:
```diff
@@ -62,6 +63,19 @@ def get_time_display(dtime, format_string=None, coerce_tz=None):
+    Returns a formatted time zone (e.g. 'Asia/Tokyo (JST +0900)') for user account settings time zone drop down
+    """
+    abbr = now().astimezone(timezone(time_zone)).strftime('%Z')
+    offset = now().astimezone(timezone(time_zone)).strftime('%z')
```
Can we DRY up the top 2 lines?

nasthagiri commented on common/djangoapps/util/date_utils.py:
```diff
@@ -89,7 +103,7 @@ def from_timestamp(timestamp):
     """
     try:
-        return datetime.utcfromtimestamp(int(timestamp)).replace(tzinfo=UTC)
+        return datetime.utcfromtimestamp(int(timestamp)).replace(tzinfo=utc)
```
Curious: What's the difference between `utc` and `UTC`?

kevinjkim replied:
@jcdyer correct me if i'm wrong. `UTC` wasn't included in the list of attributes/functions you should call, but `utc` was.

jcdyer replied:
Correct.  `django.utils.timezone.UTC` is a minimal version that gets used if pytz isn't importable.  `django.utils.timezone.utc` points to either `pytz.utc` or `django.utils.timezone.UTC`, and is recommended for general use.  pytz has a `utc` and a `UTC`, which are the same object, but pytz.utc is included in the module's `__all__`, and `UTC` isn't, indicating that `utc` is the version intended for general use.

nasthagiri commented on lms/static/js/student_account/views/account_settings_fields.js:
```diff
@@ -41,10 +41,13 @@
                 fieldTemplate: field_text_account_template,
                 successMessage: function () {
-                    return this.indicators.success + StringUtils.interpolate(
+                    return HtmlUtils.joinHtml(
```
Weren't the safe template updates merged as part of a separate PR?

jcdyer replied:
Those merged into this branch.

nasthagiri commented on lms/static/js/student_account/views/account_settings_view.js:
```diff
@@ -67,7 +68,9 @@
                 _.each(view.$('.account-settings-section-body'), function (sectionEl, index) {
                     _.each(view.options.tabSections[view.activeTab][index].fields, function (field) {
-                        $(sectionEl).append(field.view.render().el);
+                        if (field.view.enabled) {
```
Curious: Is the plan to keep the "enabled" field around even after the feature flag is removed?  Or would this code also be removed with the feature flag?

kevinjkim replied:
I'm fine with removing it after if you guys want to

nasthagiri commented on openedx/core/djangoapps/user_api/models.py:
```diff
@@ -27,6 +30,10 @@ class UserPreference(models.Model):
     key = models.CharField(max_length=255, db_index=True, validators=[RegexValidator(KEY_REGEX)])
     value = models.TextField()
 
+    TIME_ZONE_CHOICES = [
```
Why is this constant defined in the user api model?  It seems generic and basic enough to be in a more common place, right?

kevinjkim replied:
I put it there initially mainly because the options for the student model fields were put in the student model file so i put the time zone options in the user preference file. I think I'll end up moving this to the `openedx.core.lib.time_zone_utils.py` file I'll create later in the other ticket.

nasthagiri replied:
Sounds good.

kevinjkim commented:
ready for a second pass whenever you have the time

nasthagiri commented:
👍 lgtm.

jcdyer commented on common/djangoapps/util/date_utils.py:
```diff
@@ -52,7 +53,7 @@ def get_time_display(dtime, format_string=None, coerce_tz=None):
         except UnknownTimeZoneError:
-            to_tz = UTC
+            to_tz = utc
         dtime = to_tz.normalize(dtime.astimezone(to_tz))
```
I know you didn't write it, but this line looks problematic.  I need to delve back into the pytz docs to figure out what this should look like, but I think there's a problem with using both astimezone as normalize with the same tzinfo object (`to_tz`).  

Also, normalizing a datetime that isn't in UTC will raise an exception during ambiguous times (an hour before or after daylight savings stops or starts), meaning twice a year, this could potentially crash our app for users.

jcdyer replied:
Correction: I checked the pytz docs; normalizing won't cause any problems.  It's just unnecessary in this case, because no timedelta math has been performed on the datetime object after converting it to `to_tz`.

This code is slightly inefficient but harmless.  Feel free to leave it alone.

jcdyer replied:
The ambiguity exception issue occurs when creating timezone objects from scratch, not when normalizing.

jcdyer commented on lms/djangoapps/student_account/test/test_views.py:
```diff
@@ -18,22 +18,23 @@
 from third_party_auth.tests.testutil import simulate_running_pipeline, ThirdPartyAuthTestMixin
 from util.testing import UrlResetMixin
 from xmodule.modulestore.tests.django_utils import ModuleStoreTestCase
-from openedx.core.djangoapps.theming.tests.test_util import with_edx_domain_context
```
Thanks.  This looks better.

jcdyer commented:
Pass complete.  A couple relatively minor comments.

jcdyer commented:
The sorting fix will go into the new PR.  :+1:

jcdyer commented:
@sstack22 We decided not to internationalize the timezone strings, right?

jcdyer commented:
@catong Are you satisfied with the in-app messaging in this PR?

catong commented on openedx/core/djangoapps/user_api/preferences/api.py:
```diff
@@ -392,6 +394,17 @@ def validate_user_preference_serializer(serializer, preference_key, preference_v
         })
+    if preference_key == "time_zone" and preference_value not in common_timezones_set:
+        developer_message = ugettext_noop(u"Value '{preference_value}' not valid for preference '{preference_key}': Not in timezone set.")  # pylint: disable=line-too-long
+        user_message = ugettext_noop(u"Value '{preference_value}' is not valid for user preference '{preference_key}'.")
```
Could the user message be something simpler, such as " ```{preference_value} is not a valid time zone selection.```?

nasthagiri replied:
@kevinjkim @jcdyer FYI - if the doc is still being decided, feel free to create a separate PR to update the Doc.  I'd like to close out this long-term branch and make further changes in subsequent smaller PRs, especially since it's behind a feature flag.

kevinjkim replied:
@catong I changed the message to be what you specified above. I'll push the updated code in PR #12870 so that we can close this one.

catong replied:
@kevinjkim Great; thanks!

kevinjkim commented:
jenkins run bokchoy
